### PR TITLE
FIX: Index is out of bounds

### LINKED ISSUE #3014: Index is out of bounds
Requested in #3000.

```fortran
program main
integer, parameter :: n = 2
complex ::  sum
real, parameter :: d(0 : n) = [2.48574089138753566e-5,1.05142378581721974, -3.45687097222016235]
! sum = cmplx(d(0), kind = 8)
print *, d(0)
end program 
```

> _Note: this is note an issue in cmplx_

```console
% gfortran c.f90 && ./a.out
   2.48574088E-05
% lfortran c.f90 
semantic error: The index is out of bounds
 --> c.f90:6:12
  |
6 | print *, d(0)
  |            ^ 


Note: Please report unclear or confusing messages as bugs at
https://github.com/lfortran/lfortran/issues.
```

### PR DESCRIPTION
**Pull request by Pranavchiku**
Fixes #3014, towards #3000.

### CODE REVIEW

File: src/lfortran/semantics/ast_common_visitor.h [hunk 1]
```diff
@@ -3123,6 +3145,10 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
                             ASR::ArrayConstant_t *val2 = ASR::down_cast<ASR::ArrayConstant_t>(val);
                             ASR::IntegerConstant_t *index2 = ASR::down_cast<ASR::IntegerConstant_t>(index);
                             int64_t index3 = index2->m_n-1;
+                            bool is_zero_based_indexing = check_zero_based_indexing(v);
+                            if (index2->m_n == 0 && is_zero_based_indexing) {
+                                index3 = 0;
```
Comment (certik) [inline src/lfortran/semantics/ast_common_visitor.h:3150]: This seems like a quick fix for zero based indexing. But what about 5 based indexing?

Let's add a test to also start from some other number, and ensure the code above can handle it also.

Comment (Pranavchiku) [reply src/lfortran/semantics/ast_common_visitor.h:3150]: Yes, this do not work 
```fortran
program parameter_06
integer, parameter :: n = 2
complex ::  sum
real, parameter :: d(3 : n+3) = [2.48574089138753566e-5,1.05142378581721974, -3.45687097222016235]
print *, d(4)
end program 
```

```console
% gfortran a.f90 && ./a.out
   1.05142379    
% lfortran a.f90 
semantic error: The index is out of bounds
 --> a.f90:5:12
  |
5 | print *, d(4)
  |            ^ 


Note: Please report unclear or confusing messages as bugs at
https://github.com/lfortran/lfortran/issues.

```

Comment (Pranavchiku) [reply src/lfortran/semantics/ast_common_visitor.h]: Resolved with latest push.

**Review by certik — APPROVED** [review]
I think that this is good, thanks!

**Pull request merged**